molgenis-dataexplorer/src/main/java/org/molgenis/dataexplorer/controller/DataExplorerController.java

```diff
@@ -412,18 +410,34 @@ public void updateMenuOnDeleteEntity(@PathVariable("entityName") String entityNa
 	public void download(@RequestParam("dataRequest") String dataRequestStr, HttpServletResponse response)
 			throws IOException
 	{
+		DataExplorerDownloadHandler download = new DataExplorerDownloadHandler(dataService);
+
 		// Workaround because binding with @RequestBody is not possible:
 		// http://stackoverflow.com/a/9970672
 		dataRequestStr = URLDecoder.decode(dataRequestStr, "UTF-8");
 		LOG.info("Download request: [" + dataRequestStr + "]");
 		DataRequest dataRequest = new GsonHttpMessageConverter().getGson().fromJson(dataRequestStr, DataRequest.class);
 
-		String entityName = dataRequest.getEntityName();
-		String fileName = entityName + '_' + new SimpleDateFormat("yyyy-MM-dd hh:mm:ss").format(new Date()) + ".csv";
+		if (dataRequest.getDownloadType() == DownloadType.DOWNLOAD_TYPE_CSV)
+		{
+			response.setContentType("text/csv");
+			String fileName = dataRequest.getEntityName() + '_'
+					+ new SimpleDateFormat("yyyy-MM-dd hh:mm:ss").format(new Date()) + ".csv";
+			response.addHeader("Content-Disposition", "attachment; filename=" + fileName);
+
+			ServletOutputStream outputStream = response.getOutputStream();
+			download.writeToCsv(dataRequest, outputStream, ',');
+		}
+		else if (dataRequest.getDownloadType() == DownloadType.DOWNLOAD_TYPE_XLSX)
+		{
+			response.setContentType("application/vnd.openxmlformats-officedocument.spreadsheetml.sheet");
```

Comment:
Spring has content type negotiation and we probably can use that here. The idea is that you let the controller return a model object that contains all the information that should be rendered. Then Spring determines which serializer to use based on the requested content type in the header or in the extension of the request URL.